python/ray/serve/router.py

```diff
@@ -55,148 +57,158 @@ def __init__(
             self,
             controller_handle,
             backend_tag,
-            event_loop: asyncio.AbstractEventLoop,
     ):
         self.backend_tag = backend_tag
+
+        # Lock used to serialize the callback updates and assign_replica.
+        self.lock = threading.RLock()
         # NOTE(simon): We have to do this because max_concurrent_queries
         # and the replica handles come from different long poll keys.
         self.max_concurrent_queries: int = 8
+        # Track the on-going object refs assigned to replicas.
         self.in_flight_queries: Dict[ActorHandle, set] = dict()
         # The iterator used for load balancing among replicas. Using itertools
         # cycle, we implements a round-robin policy, skipping overloaded
-        # replicas.
-        # NOTE(simon): We can make this more pluggable and consider different
+        # replicas. We can make this more pluggable and consider different
         # policies like: min load, pick min of two replicas, pick replicas on
         # the same node.
         self.replica_iterator = itertools.cycle(self.in_flight_queries.keys())
 
-        # Used to unblock this replica set waiting for free replicas. A newly
-        # added replica or updated max_concurrent_queries value means the
-        # query that waits on a free replica might be unblocked on.
-        self.config_updated_event = asyncio.Event(loop=event_loop)
-        self.num_queued_queries = 0
+        self.pending_queue: List[Tuple[Query, concurrent.futures.Future]] = []
+
         self.num_queued_queries_gauge = metrics.Gauge(
             "serve_backend_queued_queries",
             description=(
                 "The current number of queries to this backend waiting"
                 " to be assigned to a replica."),
-            tag_keys=("backend", "endpoint"))
+            tag_keys=("backend", ))
         self.num_queued_queries_gauge.set_default_tags({
             "backend": self.backend_tag
         })
 
         self.long_poll_client = LongPollClient(
-            controller_handle,
-            {
+            controller_handle, {
                 (LongPollNamespace.BACKEND_CONFIGS, backend_tag): self.
                 set_max_concurrent_queries,
                 (LongPollNamespace.REPLICA_HANDLES, backend_tag): self.
                 update_worker_replicas,
-            },
-            call_in_event_loop=event_loop,
-        )
+            })
 
     def set_max_concurrent_queries(self, backend_config: BackendConfig):
         new_value: int = backend_config.max_concurrent_queries
         if new_value != self.max_concurrent_queries:
             self.max_concurrent_queries = new_value
             logger.debug(
                 f"ReplicaSet: changing max_concurrent_queries to {new_value}")
-            self.config_updated_event.set()
+        self._drain_pending_queue()
```

Comment:
I don't fully understand why we are draining the pending queue here. Is this required for the config change to take place?

Reply:
Please add a comment explaining this as well.

Reply:
I see now that this isn't actually draining the queries, it's just trying to send them. Makes sense to me now.